Pass namespaced service name to cloudprovider's EnsureLoadBalancer

therc opened the pull request:
Also has an AWS implementation that plugs the service name into the ELB and SG.
Fixes #20668 
cc @justinsb @thockin

Commit:
Pass namespaced service name to cloudprovider's EnsureLoadBalancer

Also has an AWS implementation that plugs the service name into the ELB and SG.
Log the service name under GCE and OpenStack.
Fixes #20668

therc commented:
Rebased!

justinsb commented on pkg/cloudprovider/cloud.go:
```diff
@@ -83,7 +84,7 @@ type LoadBalancer interface {
 	GetLoadBalancer(name, region string) (status *api.LoadBalancerStatus, exists bool, err error)
 	// EnsureLoadBalancer creates a new load balancer, or updates an existing one. Returns the status of the balancer
-	EnsureLoadBalancer(name, region string, loadBalancerIP net.IP, ports []*api.ServicePort, hosts []string, affinityType api.ServiceAffinity) (*api.LoadBalancerStatus, error)
+	EnsureLoadBalancer(namespacedName types.NamespacedName, name, region string, loadBalancerIP net.IP, ports []*api.ServicePort, hosts []string, affinityType api.ServiceAffinity) (*api.LoadBalancerStatus, error)
```
suggest renaming namespacedName to serviceName so that it is clear what the name refers to.  Also it's now ambiguous as to whether we are creating/updating based on namespacedName vs name, so I suggest we expand the comment to specify, and maybe move namespacedName towards the end of the arguments (first place is misleading as to its importance)

therc replied:
Done all three.

justinsb commented on pkg/cloudprovider/providers/aws/aws_loadbalancer.go:
```diff
@@ -47,7 +48,12 @@ func (s *AWSCloud) ensureLoadBalancer(name string, listeners []*elb.Listener, su
 
-		glog.Info("Creating load balancer with name: ", name)
+		createRequest.Tags = []*elb.Tag{
+			{Key: aws.String(TagNameKubernetesCluster), Value: aws.String(s.getClusterName())},
```
Ah ... good catch.  Including the cluster tag is a good thing to do.  ELB isn't quite as good at letting us actually use these tags easily, but we should still tag it.

justinsb commented on pkg/cloudprovider/providers/aws/aws_loadbalancer.go:
```diff
@@ -47,7 +48,12 @@ func (s *AWSCloud) ensureLoadBalancer(name string, listeners []*elb.Listener, su
-		glog.Info("Creating load balancer with name: ", name)
+		createRequest.Tags = []*elb.Tag{
+			{Key: aws.String(TagNameKubernetesCluster), Value: aws.String(s.getClusterName())},
+			{Key: aws.String("KubernetesService"), Value: aws.String(namespacedName.String())},
```
I think we should keep tag name in the kubernetes.io namespace (I screwed up on the format of TagNameKubernetesCluster;).  How about `kubernetes.io/service-name`?  (And I suggest making it a constant so we group these magic tags together).

therc replied:
Done both.

justinsb commented:
A few nits, but AWS side looks pretty good.  And other than logging, this is only used on AWS at the moment, right @therc ?

therc commented:
Yes, at the moment only the AWS cloud provider is doing anything useful with the information.
I wouldn't know where to start (and how to test) with the other two.

Incorporated all feedback, rebased and squashed.

justinsb commented:
LGTM - thanks for fixing. @thockin #20668 was your issue - what do you think?

thockin commented:
Thanks!  Do you want to do the GCE work, or just get this in an one of us will do it eventually?  I'm happy to point you at the right docs.

thockin commented:
add to whitelist

therc commented:
I tried to do the GCE work based on documentation alone in #20909 and that hasn't turned out so well. Unless it's a flake...
